driver-core/src/main/com/mongodb/internal/operation/ClientBulkWriteOperation.java

```diff
@@ -180,23 +194,79 @@ public ClientBulkWriteResult execute(final WriteBinding binding) throws ClientBu
+            });
+        }).<ClientBulkWriteResult>thenApply((transformedTopLevelError, callback1) -> {
+            callback1.onResult(resultAccumulator.build(transformedTopLevelError, effectiveWriteConcern), null);
+        }).finish(finalCallback);
+    }
+
     /**
      * To execute a batch means:
      * <ul>
      *     <li>execute a `bulkWrite` command, which creates a cursor;</li>
      *     <li>consume the cursor, which may involve executing `getMore` commands.</li>
      * </ul>
      *
-     * @throws MongoException When a {@linkplain ClientBulkWriteException#getError() top-level error} happens.
+     * @throws MongoException When a {@linkplain ClientBulkWriteException#getCause() top-level error} happens.
      */
     private void executeAllBatches(
             final WriteConcern effectiveWriteConcern,
             final WriteBinding binding,
             final ResultAccumulator resultAccumulator) throws MongoException {
-        Integer nextBatchStartModelIndex = 0;
+        Integer nextBatchStartModelIndex = INITIAL_BATCH_MODEL_START_INDEX;
         do {
             nextBatchStartModelIndex = executeBatch(nextBatchStartModelIndex, effectiveWriteConcern, binding, resultAccumulator);
-        } while (nextBatchStartModelIndex != null);
+        } while (shouldExecuteNextBatch(nextBatchStartModelIndex));
+    }
+
+    /**
+     * To execute a batch means:
+     * <ul>
+     *     <li>execute a `bulkWrite` command, which creates a cursor;</li>
+     *     <li>consume the cursor, which may involve executing `getMore` commands.</li>
+     * </ul>
+     *
+     * {@link SingleResultCallback<Void>} propagates {@link MongoException} when a {@linkplain ClientBulkWriteException#getCause() top-level error} happens.
+     */
+    private void executeAllBatchesAsync(
+            final WriteConcern effectiveWriteConcern,
+            final AsyncWriteBinding binding,
+            final ResultAccumulator resultAccumulator,
+            final SingleResultCallback<Void> finalCallback) {
+        MutableValue<Integer> nextBatchStartModelIndex = new MutableValue<>(INITIAL_BATCH_MODEL_START_INDEX);
+
+        beginAsync()
+                .thenRunDoWhileLoop(iterationCallback -> {
+                    beginAsync().<Integer>thenSupply(c -> {
+                        executeBatchAsync(nextBatchStartModelIndex.get(), effectiveWriteConcern, binding, resultAccumulator, c);
+                    }).<Void>thenApply((nextBatchStartModelIdx, c) -> {
+                        nextBatchStartModelIndex.set(nextBatchStartModelIdx);
+                        c.complete(c);
+                    }).finish(iterationCallback);
+                }, () -> shouldExecuteNextBatch(nextBatchStartModelIndex.getNullable()))
+                .finish(finalCallback);
+    }
+
+    private static boolean shouldExecuteNextBatch(@Nullable final Integer nextBatchStartModelIndex) {
+        return nextBatchStartModelIndex != null;
```

Comment:
[optional]

Given how trivial this method, I find the code more readable when the check is done explicitly, instead of calling this method: I tried the change, and the `nextBatchStartModelIndex != null`/`nextBatchStartModelIndex.getNullable() != null` checks are immediately obvious, while with `shouldExecuteNextBatch` I have to navigate to the method in order to understand the condition.

Maybe we should inline the method?